{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/glm5uh/anaconda3/envs/driven_devel/lib/python3.7/site-packages/ipykernel_launcher.py:40 \u001b[1;31mFutureWarning\u001b[0m: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n",
      "/Users/glm5uh/anaconda3/envs/driven_devel/lib/python3.7/site-packages/ipykernel_launcher.py:80 \u001b[1;31mSettingWithCopyWarning\u001b[0m: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import cobra\n",
    "import driven\n",
    "import numpy as np\n",
    "from driven.data_sets import ExpressionProfile\n",
    "from pickle import load\n",
    "import json\n",
    "import matplotlib\n",
    "\n",
    "fname = '../data/nobori_2018/pnas.1800529115.sd02.xlsx'\n",
    "conditions = ['MM','vivo','flg22','D36E','AvrRpt2','AvrRps4']\n",
    "all_conditions_frames = []\n",
    "condition_to_p = {}\n",
    "# load the up and down sheet for each comparison, then merge\n",
    "# them into a single dataframe and rename the columns to meet the\n",
    "# conventions in the driven package.\n",
    "for condition in conditions:\n",
    "    down = pd.read_excel(fname,sheet_name=condition+\"_down\")\n",
    "    down = down.rename({condition+'_down':'gene'},axis='columns')\n",
    "    up = pd.read_excel(fname,sheet_name=condition+\"_up\")\n",
    "    up = up.rename({condition+'_up':'gene'},axis='columns')\n",
    "    both = pd.concat([down,up])\n",
    "    fchange_col = both.columns\n",
    "    expression_column = [c for c in both.columns if \"fold change\" in c][0]\n",
    "    p_column = [c for c in both.columns if \"q-value\" in c][0]\n",
    "    condition = expression_column.split(': ')[1]\n",
    "    \n",
    "    both = both.rename({p_column:\\\n",
    "        \"p-value-\"+condition,expression_column:condition},axis='columns')\n",
    "    \n",
    "    # split each gene name by '-' to get the KEGG gene ID only,\n",
    "    # then reassign it to the index\n",
    "    new_index = [s[0] for s in both['gene'].str.split('-')]\n",
    "    both.index = new_index\n",
    "    both = both.drop('gene',axis=1)\n",
    "    all_conditions_frames.append(both)\n",
    "    \n",
    "    # save the exact condition name and the p value column associated with it\n",
    "    condition_to_p[condition] = \"p-value-\"+condition\n",
    "xomics_data = pd.concat(all_conditions_frames,axis=1)\n",
    "\n",
    "# for p-value columns that are NaN after the merge, reassign as 1 to indicate\n",
    "# not significant\n",
    "p_cols = [c for c in xomics_data.columns if \"p-value\" in c]\n",
    "xomics_data[p_cols] = xomics_data[p_cols].fillna(value=1.0)\n",
    "\n",
    "# for expression, do the equivalent (set foldchange to 0)\n",
    "diff_cols = [c for c in xomics_data.columns if \"p-value\" not in c]\n",
    "xomics_data[diff_cols] = xomics_data[diff_cols].fillna(value=0.0)\n",
    "\n",
    "\n",
    "# Load the count data\n",
    "fname = \"../data/nobori_2018/GSE103441_count.txt\"\n",
    "count_frame = pd.read_csv(fname,sep='\\t',index_col=0)\n",
    "# remove sample names from the end of column names\n",
    "# and merge samples group-wise into a mean\n",
    "#'_'.join(count_frame.columns[0].split('_')[:-1])\n",
    "groups = set(['_'.join(c.split('_')[:-1]) for c in count_frame.columns])\n",
    "mean_count_frame = pd.DataFrame(index=count_frame.index)\n",
    "group_to_samples = {}\n",
    "for group in groups:\n",
    "    samps = [col for col in count_frame.columns if group in col]\n",
    "    group_to_samples[group] = samps\n",
    "    # normalize to transcripts per million within each sample\n",
    "    mean = count_frame[samps].divide(count_frame[samps].sum()/1E6).mean(axis=1)\n",
    "    mean_count_frame[group] = mean\n",
    "    \n",
    "    \n",
    "# extract the groups of interest from count frame\n",
    "flg_v_mock_counts = mean_count_frame[['Pto_col_mock','Pto_col_flg22']]\n",
    "flg_v_mm_counts = mean_count_frame[[\"Pto_col_flg22\", \"Pto_MM_A80\"]]\n",
    "# replace A in the gene names with _ as in the other\n",
    "# spreadsheets provided by the author\n",
    "#the following line is not necessary because PSPTO_0001 and PSPTOA0001 are different genes, changed to \n",
    "#replace \"A\" with \"_A\" like plasmid B\n",
    "flg_v_mock_counts.index = flg_v_mock_counts.index.str.replace('A','_A')\n",
    "flg_v_mm_counts.index = flg_v_mm_counts.index.str.replace('A', \"_A\")\n",
    "# add p-values\n",
    "flg_v_mock_p = xomics_data[\"p-value-flg22_Pto - Mock_Pto\"]\n",
    "flg_v_mock_counts['p-value'] = flg_v_mock_p\n",
    "flg_v_mock_counts = flg_v_mock_counts.fillna(1)\n",
    "\n",
    "# rename the gene identifiers by splitting at \"-\" to only include the actual identifiers,\n",
    "# not the descriptions.\n",
    "identifiers = [gene.split('-')[0] for gene in flg_v_mock_counts.index.tolist()]\n",
    "\n",
    "flg_v_mock_exp_prof = ExpressionProfile(identifiers=identifiers,\n",
    "                                        conditions=['Pto_col_mock','Pto_col_flg22'],\n",
    "                                        expression=flg_v_mock_counts[['Pto_col_mock','Pto_col_flg22']].values,\n",
    "                                        p_values=np.reshape(flg_v_mock_counts['p-value'].values,(-1, 1)))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load the medusa ensemble\n",
    "import medusa\n",
    "with open('../results/ensembles/psy_ensemble_100_for_analysis.pickle', 'rb') as infile:\n",
    "    ensemble = load(infile)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>PST</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x0126813da0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>1224</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>1439</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>58</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*bio1 - 1.0*bio1_reverse_b18f7</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>c, e</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model PST at 0x126813da0>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensemble.base_model\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Pto_col_mock                          297.386577\n",
       "Pto_col_flg22                         265.865424\n",
       "Pto_col_mock Pto_col_flg22 p-value      1.000000\n",
       "dtype: float64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXcAAAD4CAYAAAAXUaZHAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAATNElEQVR4nO3df6zddX3H8eebVotyO9uKXDrarGWrZsVEsDeMjmW5V0Q6RyjLJLlmui7D9I9h4tzMbEeyxD8adFt0WZBpY92aFb12FUZhQ9fU3i1LgEpVhAK1lbJSC1SxMC9LOsve++N8Sw+35/aee+8595zz4flIbs73+/n+ep1L7+t87/d87yEyE0lSWc7rdABJUutZ7pJUIMtdkgpkuUtSgSx3SSrQ3E4HALjwwgtz2bJl097+5Zdf5oILLmhdoDbplZzQO1l7JSf0TlZztl67su7bt+8nmfm2hgszs+Nfq1atypnYs2fPjLafLb2SM7N3svZKzszeyWrO1mtXVuDhnKBXvSwjSQWy3CWpQJa7JBXIcpekAlnuklQgy12SCmS5S1KBLHdJKpDlLkkFKqPcf/Yc7Lmt9iVJKqTcJUmvYblLUoEsd0kqkOUuSQWy3CWpQJa7JBXIcpekAlnuklQgy12SCmS5S1KBLHdJKpDlLkkFstwlqUCWuyQVyHKXpAJZ7pJUIMtdkgpkuUtSgSx3SSqQ5S5JBWqq3CPi6Yh4NCK+FxEPV2OLImJXRBysHhfWrb8xIg5FxIGIuK5d4SVJjU3lzH0oMy/PzIFqfgOwOzNXALureSJiJTAMXAasAe6IiDktzCxJmsRMLsusBbZW01uBG+vGRzLzZGYeBg4BV87gOJKkKYrMnHyliMPACSCBL2bm5oh4MTMX1K1zIjMXRsTtwIOZua0a3wLcn5k7xu1zPbAeoL+/f9XIyMi0n8TYSyfoO+9kbWb+xdPeT7uNjY3R19fX6RhN6ZWsvZITeierOVuvXVmHhob21V1NeY25Te7j6sw8FhEXAbsi4slzrBsNxs56BcnMzcBmgIGBgRwcHGwyytlG7x1hsO9wbWZweNr7abfR0VFm8jxnU69k7ZWc0DtZzdl6ncja1GWZzDxWPR4H7qZ2meX5iFgMUD0er1Y/Ciyt23wJcKxVgSVJk5u03CPigoiYf3oaeB/wGLATWFettg64p5reCQxHxLyIWA6sAPa2OrgkaWLNXJbpB+6OiNPrfyUzvxER3wa2R8TNwBHgJoDM3B8R24HHgVPALZn5SlvSS5IamrTcM/Mp4F0Nxl8Arplgm03AphmnkyRNi3+hKkkFstwlqUCWuyQVyHKXpAJZ7pJUIMtdkgpkuUtSgSx3SSqQ5S5JBbLcJalAlrskFchyl6QCWe6SVCDLXZIKZLlLUoEsd0kqkOUuSQWy3CWpQJa7JBXIcpekAlnuklQgy12SCmS5S1KBLHdJKpDlLkkFstwlqUCWuyQVqOlyj4g5EfHdiLivml8UEbsi4mD1uLBu3Y0RcSgiDkTEde0ILkma2FTO3D8GPFE3vwHYnZkrgN3VPBGxEhgGLgPWAHdExJzWxJUkNaOpco+IJcBvA1+qG14LbK2mtwI31o2PZObJzDwMHAKubE1cSVIzIjMnXyliB3AbMB/4RGZeHxEvZuaCunVOZObCiLgdeDAzt1XjW4D7M3PHuH2uB9YD9Pf3rxoZGZn2kxh76QR9552szcy/eNr7abexsTH6+vo6HaMpvZK1V3JC72Q1Z+u1K+vQ0NC+zBxotGzuZBtHxPXA8czcFxGDTRwvGoyd9QqSmZuBzQADAwM5ONjMrhsbvXeEwb7DtZnB4Wnvp91GR0eZyfOcTb2StVdyQu9kNWfrdSLrpOUOXA3cEBHvB84HfiEitgHPR8TizHw2IhYDx6v1jwJL67ZfAhxrZWhJ0rlNes09Mzdm5pLMXEbtjdJvZeaHgJ3Aumq1dcA91fROYDgi5kXEcmAFsLflySVJE2rmzH0inwa2R8TNwBHgJoDM3B8R24HHgVPALZn5yoyTSpKaNqVyz8xRYLSafgG4ZoL1NgGbZphNkjRN/oWqJBXIcpekAlnuklQgy12SCmS5S1KBLHdJKpDlLkkFstwlqUCWuyQVyHKXpAJZ7pJUIMtdkgpkuUtSgSx3SSqQ5S5JBbLcJalAlrskFchyl6QCWe6SVCDLXZIKZLlLUoEsd0kqkOUuSQWy3CWpQJa7JBXIcpekAlnuklQgy12SCjRpuUfE+RGxNyIeiYj9EfGpanxRROyKiIPV48K6bTZGxKGIOBAR17XzCUiSztbMmftJ4D2Z+S7gcmBNRFwFbAB2Z+YKYHc1T0SsBIaBy4A1wB0RMacd4SVJjU1a7lkzVs2+ofpKYC2wtRrfCtxYTa8FRjLzZGYeBg4BV7Y0tSTpnCIzJ1+pdua9D/gV4POZ+cmIeDEzF9StcyIzF0bE7cCDmbmtGt8C3J+ZO8btcz2wHqC/v3/VyMjItJ/E2Esn6DvvZG1m/sXT3k+7jY2N0dfX1+kYTemVrL2SE3onqzlbr11Zh4aG9mXmQKNlc5vZQWa+AlweEQuAuyPinedYPRrtosE+NwObAQYGBnJwcLCZKA2N3jvCYN/h2szg8LT3026jo6PM5HnOpl7J2is5oXeymrP1OpF1SnfLZOaLwCi1a+nPR8RigOrxeLXaUWBp3WZLgGMzTipJalozd8u8rTpjJyLeBLwXeBLYCayrVlsH3FNN7wSGI2JeRCwHVgB7Wx1ckjSxZi7LLAa2VtfdzwO2Z+Z9EfEAsD0ibgaOADcBZOb+iNgOPA6cAm6pLutIkmbJpOWemd8Hrmgw/gJwzQTbbAI2zTidJGla/AtVSSqQ5S5JBbLcJalAlrskFchyl6QCWe6SVCDLXZIKZLlLUoEsd0kqkOUuSQWy3CWpQJa7JBXIcpekAlnuklQgy12SCmS5S1KBLHdJKpDlLkkFstwlqUCWuyQVyHKXpAJZ7pJUIMtdkgpkuUtSgSx3SSqQ5S5JBbLcJalAk5Z7RCyNiD0R8URE7I+Ij1XjiyJiV0QcrB4X1m2zMSIORcSBiLiunU9AknS2Zs7cTwF/mpm/ClwF3BIRK4ENwO7MXAHsruaplg0DlwFrgDsiYk47wkuSGpu03DPz2cz8TjX9M+AJ4BJgLbC1Wm0rcGM1vRYYycyTmXkYOARc2ergkqSJTemae0QsA64AHgL6M/NZqL0AABdVq10CPFO32dFqTJI0SyIzm1sxog/4d2BTZt4VES9m5oK65Scyc2FEfB54IDO3VeNbgH/NzK+P2996YD1Af3//qpGRkWk/ibGXTtB33snazPyLp72fdhsbG6Ovr6/TMZrSK1l7JSf0TlZztl67sg4NDe3LzIFGy+Y2s4OIeAPwdeDOzLyrGn4+IhZn5rMRsRg4Xo0fBZbWbb4EODZ+n5m5GdgMMDAwkIODg81EaWj03hEG+w7XZgaHp72fdhsdHWUmz3M29UrWXskJvZPVnK3XiazN3C0TwBbgicz8bN2incC6anodcE/d+HBEzIuI5cAKYG/rIkuSJtPMmfvVwIeBRyPie9XYnwOfBrZHxM3AEeAmgMzcHxHbgcep3WlzS2a+0vLkkqQJTVrumfmfQEyw+JoJttkEbJpBLknSDPgXqpJUIMtdkgpkuUtSgSx3SSqQ5S5JBbLcJalAlrskFchyl6QCWe6SVCDLXZIKZLlLUoEsd0kqkOUuSQWy3CWpQJa7JBWoiHJ/+eSpTkeQpK5SRLlLkl7LcpekAlnuklQgy12SCmS5S1KBLHdJKtDcTgdouT23nZke2ti5HJLUQZ65S1KBLHdJKpDlLkkFstwlqUCWuyQVaNJyj4gvR8TxiHisbmxRROyKiIPV48K6ZRsj4lBEHIiI69oVXJI0sWbO3P8BWDNubAOwOzNXALureSJiJTAMXFZtc0dEzGlZWklSUyYt98z8D+Cn44bXAlur6a3AjXXjI5l5MjMPA4eAK1uUVZLUpMjMyVeKWAbcl5nvrOZfzMwFdctPZObCiLgdeDAzt1XjW4D7M3NHg32uB9YD9Pf3rxoZGZn2k3jpxE95y9z/PXvB/Iunvc92GBsbo6+vr9MxmtIrWXslJ/ROVnO2XruyDg0N7cvMgUbLWv0XqtFgrOGrR2ZuBjYDDAwM5ODg4LQP+i87tjG44JmzFwwOT3uf7TA6OspMnuds6pWsvZITeierOVuvE1mne7fM8xGxGKB6PF6NHwWW1q23BDg2/XiSpOmYbrnvBNZV0+uAe+rGhyNiXkQsB1YAe2cWUZI0Vc3cCvlV4AHgHRFxNCJuBj4NXBsRB4Frq3kycz+wHXgc+AZwS2a+0q7w9R546oXZOIwk9YRJr7ln5gcnWHTNBOtvAjbNJJQkaWb8C1VJKpDlLkkFstwlqUCWuyQVyHKXpAIVVe7eDilJNUWVuySpxnKXpAJZ7pJUIMtdkgpkuUtSgSx3SSqQ5S5JBWr1/4mpu+y57cz00MbO5ZCkWVbcmbt/yCRJBZa7JMlyl6QiWe6SVCDLXZIKVPbdMvW8c0bS60iRZ+4PPPWCd81Iel0r+sz9gadeYPWlbz17gWfxkgpX5Jm7JL3eWe6SVKDiy91r75Jej4ovd0l6PSr6DdXT6s/ez3qDtf7N1Xq+0Sqph70uyn28Ce+imYh310jqMW27LBMRayLiQEQciogN7TrOdNWfzU/puvye2858SVKXasuZe0TMAT4PXAscBb4dETsz8/F2HG8qxhf56fmzzuanWt4Tnd3Xj8fqqW0rSdPUrssyVwKHMvMpgIgYAdYCHS/3ehMVfbNWc9ur26y+9K2vvkA8sOUTr87XL+Oi52DPbWfWG/d4Vr5qPwCfO/W7fPzat/O5XT/g49e+/TXrnD4G8JoXh9Prfm7XD7jqyOYz69WtU3+Mc72wjD/uWZp5cZvqC9dUtz3XC/JUt2/3i6wv6K9vs/DfPzKz9TuN+ACwJjM/Us1/GPi1zPxo3TrrgfXV7DuAAzM45IXAT2aw/WzplZzQO1l7JSf0TlZztl67sv5SZr6t0YJ2nblHg7HXvIpk5mZgc0sOFvFwZg60Yl/t1Cs5oXey9kpO6J2s5my9TmRt1xuqR4GldfNLgGNtOpYkaZx2lfu3gRURsTwi3ggMAzvbdCxJ0jhtuSyTmaci4qPAN4E5wJczc387jlVpyeWdWdArOaF3svZKTuidrOZsvVnP2pY3VCVJneVny0hSgSx3SSpQT5d7pz/iICKWRsSeiHgiIvZHxMeq8UURsSsiDlaPC+u22VjlPRAR19WNr4qIR6tlfxsRjW4nnWneORHx3Yi4r8tzLoiIHRHxZPW9Xd2NWSPi49V/98ci4qsRcX635IyIL0fE8Yh4rG6sZdkiYl5EfK0afygilrU4619V//2/HxF3R8SCTmdtlLNu2SciIiPiwk7nfFVm9uQXtTdqfwhcCrwReARYOcsZFgPvrqbnAz8AVgJ/CWyoxjcAn6mmV1Y55wHLq/xzqmV7gdXU/kbgfuC32pD3T4CvAPdV892acyvwkWr6jcCCbssKXAIcBt5UzW8H/qBbcgK/CbwbeKxurGXZgD8CvlBNDwNfa3HW9wFzq+nPdEPWRjmr8aXUbh75L+DCTud8NVerfzBn66v65nyzbn4jsLHDme6h9nk6B4DF1dhi4ECjjNU/iNXVOk/WjX8Q+GKLsy0BdgPv4Uy5d2POX6BWmjFuvKuyUiv3Z4BF1O46u68qpK7JCSzjtYXZsmyn16mm51L768toVdZxy34HuLMbsjbKCewA3gU8zZly7/j3tJcvy5z+4TrtaDXWEdWvUFcADwH9mfksQPV4UbXaRJkvqabHj7fS3wB/Bvxf3Vg35rwU+DHw99UlpC9FxAXdljUzfwT8NXAEeBZ4KTP/rdtyjtPKbK9uk5mngJeAKXyO9pT8IbUz3K7LGhE3AD/KzEfGLep4zl4u90k/4mC2REQf8HXgjzPzv8+1aoOxPMd4S0TE9cDxzNzX7CYT5JmN7/lcar/6/l1mXgG8TO0SwkQ69T1dSO3D8JYDvwhcEBEfOtcmE+Tphn/H08k2K7kj4lbgFHDnJMed9awR8WbgVuAvGi2e4JizlrOXy70rPuIgIt5ArdjvzMy7quHnI2JxtXwxcLwanyjz0Wp6/HirXA3cEBFPAyPAeyJiWxfmPH3so5n5UDW/g1rZd1vW9wKHM/PHmflz4C7g17swZ71WZnt1m4iYC7wF+Gkrw0bEOuB64PeyulbRZVl/mdqL+yPVz9YS4DsRcXE35Ozlcu/4RxxU73JvAZ7IzM/WLdoJrKum11G7Fn96fLh6V3w5sALYW/2K/LOIuKra5+/XbTNjmbkxM5dk5jJq36dvZeaHui1nlfU54JmIeEc1dA21j4rutqxHgKsi4s3V/q8BnujCnPVama1+Xx+g9m+qlb8ZrQE+CdyQmf8z7jl0RdbMfDQzL8rMZdXP1lFqN1g81xU5p3uxvhu+gPdTu0Plh8CtHTj+b1D7ten7wPeqr/dTu062GzhYPS6q2+bWKu8B6u6KAAaAx6pltzODN1ImyTzImTdUuzIncDnwcPV9/WdgYTdmBT4FPFkd4x+p3RnRFTmBr1J7L+Dn1Ern5lZmA84H/gk4RO3uj0tbnPUQtevPp3+uvtDprI1yjlv+NNUbqp3+nmamHz8gSSXq5csykqQJWO6SVCDLXZIKZLlLUoEsd0kqkOUuSQWy3CWpQP8Pmhb6LExQsb4AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "\n",
    "\n",
    "# plot the distribution of transcripts for genes that are in the model\n",
    "transcripts_in_model = list(set(flg_v_mock_exp_prof.identifiers).intersection(set(\n",
    "                        [g.id for g in ensemble.base_model.genes])))\n",
    "\n",
    "\n",
    "flg_v_mock_exp_prof.data_frame.loc[transcripts_in_model]['Pto_col_flg22'].hist(bins = 100, alpha = 0.5)\n",
    "flg_v_mock_exp_prof.data_frame.loc[transcripts_in_model]['Pto_col_mock'].hist(bins = 100, alpha = 0.5)\n",
    "\n",
    "mock_mean = flg_v_mock_exp_prof.data_frame.loc[transcripts_in_model].mean()['Pto_col_mock']\n",
    "flg_mean = flg_v_mock_exp_prof.data_frame.loc[transcripts_in_model].mean()['Pto_col_flg22']\n",
    "avg2x = (mock_mean + flg_mean)*2\n",
    "\n",
    "flg_v_mock_exp_prof.data_frame.loc[transcripts_in_model].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "14\n",
      "15\n",
      "16\n",
      "17\n",
      "18\n",
      "19\n",
      "20\n",
      "21\n",
      "22\n",
      "23\n",
      "24\n",
      "25\n",
      "26\n",
      "27\n",
      "28\n",
      "29\n",
      "30\n",
      "31\n",
      "32\n",
      "33\n",
      "34\n",
      "35\n",
      "36\n",
      "37\n",
      "38\n",
      "39\n",
      "40\n",
      "41\n",
      "42\n",
      "43\n",
      "44\n",
      "45\n",
      "46\n",
      "47\n",
      "48\n",
      "49\n",
      "50\n",
      "51\n",
      "52\n",
      "53\n",
      "54\n",
      "55\n",
      "56\n",
      "57\n",
      "58\n",
      "59\n",
      "60\n",
      "61\n",
      "62\n",
      "63\n",
      "64\n",
      "65\n",
      "66\n",
      "67\n",
      "68\n",
      "69\n",
      "70\n",
      "71\n",
      "72\n",
      "73\n",
      "74\n",
      "75\n",
      "76\n",
      "77\n",
      "78\n",
      "79\n",
      "80\n",
      "81\n",
      "82\n",
      "83\n",
      "84\n",
      "85\n",
      "86\n",
      "87\n",
      "88\n",
      "89\n",
      "90\n",
      "91\n",
      "92\n",
      "93\n",
      "94\n",
      "95\n",
      "96\n",
      "97\n",
      "98\n",
      "99\n",
      "100\n"
     ]
    }
   ],
   "source": [
    "#apply gimme to all the ensemble members and save the solutions\n",
    "mock_solutions = {}\n",
    "flg_solutions = {}\n",
    "\n",
    "cutoff = avg2x\n",
    "\n",
    "member_count = 0\n",
    "for member in ensemble.members:\n",
    "    member_count += 1\n",
    "    print(member_count)\n",
    "    #if member_count is not 71: # GIMME hangs on this model. Seems to get stuck at solver; kernel interrupt doesn't work\n",
    "    ensemble.set_state(member)\n",
    "    model, mock_solutions[member.id] = driven.gimme(ensemble.base_model,flg_v_mock_exp_prof,condition = 'Pto_col_mock',\n",
    "                                                       cutoff=cutoff,fraction_of_optimum=0.05)\n",
    "    model, flg_solutions[member.id] = driven.gimme(ensemble.base_model,flg_v_mock_exp_prof,condition = 'Pto_col_flg22',\n",
    "                                                       cutoff=cutoff,fraction_of_optimum=0.05)\n",
    "\n",
    "# save the fluxes from each solution    \n",
    "pd.DataFrame([mock_solutions[member].fluxes for member in mock_solutions.keys()]).to_csv('../results/gimme/mock_gimme_tpm_fluxes.tsv',sep='\\t')\n",
    "pd.DataFrame([flg_solutions[member].fluxes for member in flg_solutions.keys()]).to_csv('../results/gimme/flg_gimme_tpm_fluxes.tsv',sep='\\t')\n",
    "\n",
    "\n",
    "# for each gimme solution, get the list of open reactions.\n",
    "def get_open_reactions(gimme_solution):\n",
    "    open_reactions = []\n",
    "    for reaction in gimme_solution.fluxes.keys():\n",
    "        if abs(gimme_solution.fluxes[reaction]) > 0:\n",
    "            open_reactions.append(reaction)\n",
    "    return open_reactions\n",
    "\n",
    "mock_open_reactions = {}\n",
    "flg_open_reactions = {}\n",
    "for member_id in mock_solutions.keys():\n",
    "    mock_open_reactions[member_id] = get_open_reactions(mock_solutions[member_id])\n",
    "    flg_open_reactions[member_id] = get_open_reactions(flg_solutions[member_id])\n",
    "    \n",
    "# save the open reactions\n",
    "with open('../results/gimme/mock_gimme_open_reactions_cutoff4.json', 'w') as outfile:\n",
    "    json.dump(mock_open_reactions,outfile)\n",
    "with open('../results/gimme/flg_gimme_open_reactions_cutoff4.json', 'w') as outfile:\n",
    "    json.dump(flg_open_reactions,outfile)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "driven_devel",
   "language": "python",
   "name": "driven_devel"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
